src/btree/bt_prefetch.c

```diff
@@ -25,10 +25,38 @@ __wt_btree_prefetch(WT_SESSION_IMPL *session, WT_REF *ref)
     conn = S2C(session);
     block_preload = 0;
 
-    if (!(F_ISSET(ref, WT_REF_FLAG_LEAF)) || (__wt_session_gen(session, WT_GEN_SPLIT) == 0))
-        return (WT_ERROR);
+    /* Don't add refs from trees that have eviction disabled since they are probably being closed */
+    if (!(F_ISSET(ref, WT_REF_FLAG_LEAF)) || S2BT(session)->evict_disabled > 0 ||
+      (__wt_session_gen(session, WT_GEN_SPLIT) == 0)) {
+        WT_STAT_CONN_INCR(session, block_prefetch_failed_start);
+        return (0);
+    }
+
+    /*
+     * We want to avoid the scenario of requesting pre-fetch on one particular ref many times (e.g
+     * when reading along a single page). We can identify this by checking if the previous pre-fetch
+     * was performed using the same home ref.
+     *
+     * In the event that we find this to be true, we perform pre-fetch for approximately the number
+     * of pages that were added to the queue (WT_PREFETCH_QUEUE_PER_TRIGGER). We then want to ensure
+     * that we will not pre-fetch from this ref for a while, and this is done by checking a counter.
+     *
+     * The counter variable prefetch_skipped_with_parent tracks the number of skips we have
+     * performed on a particular ref. If the number of skips surpasses the number of pages that have
+     * been queued for pre-fetch, we are okay to pre-fetch from this ref again. This condition will
+     * evaluate to false and the counter will be reset, effectively marking the ref as available to
+     * pre-fetch from.
+     */
+    if (session->pf.prefetch_prev_ref_home == ref->home &&
+      session->pf.prefetch_skipped_with_parent < WT_PREFETCH_QUEUE_PER_TRIGGER) {
+        ++session->pf.prefetch_skipped_with_parent;
+        WT_STAT_CONN_INCR(session, block_prefetch_skipped_same_ref);
+        WT_STAT_CONN_INCR(session, block_prefetch_skipped);
+        return (0);
+    }
```

Comment:
What was the motivation behind moving this block?

Reply:
Earlier, this check was inside the function where we pushed the ref into the queue. All the child ref have the same parent page pointer. Checking it inside and ignoring it for all the child pages is just a waste of work. Instead, check it outside and return if the same parent page has already been processed.

Reply:
Got it - thank you.